enhancements/node-pull-credentials/pull-credentials.md

```diff
@@ -0,0 +1,193 @@
+  * This should be done in the future through another enhancement proposal as
+  we currently have different sources of truth when it comes to CAs(different
+  methods of providing them to different parts of the codebase).
+
+## Proposal
+
+### User Stories
+
+As an OpenShift user
+I want to be able to import imagestreams from any image registry that the
+cluster nodes can pull images from
+So that I can use them without manually creating any extra credentials
+
+As an OpenShift user
+I want to be able to use, during builds, images from any image registry that
+the cluster nodes can pull images from
+So that I can use these images as base for my own images
+
+### Implementation Details/Notes/Constraints
+
+This is the node's filesystem path we are taking into account:
+
+- `/var/lib/kubelet/config.json` contains node's pull credentials.
+
+#### Image Stream Import
+
+- Mount the location as `hostPath` during `openshift-apiserver` deploy under
+  `/node/var/lib/kubelet/config.json`.
+
+```
+volumeMounts:
+- mountPath: /node/var/lib/kubelet/config.json
+  name: node-pull-credentials
+volumes:
+- name: node-pull-credentials
+  hostPath:
+    path: /var/lib/kubelet/config.json
+    type: File
+```
+
+- When user requires to import an image stream, `openshift-apiserver` would
+  merge all pull credentials found in `/node/var/lib/kubelet/config.json` with
+  other credentials that may exist on the namespace.
+
+#### Builds
+
+- `controller-manager` would mount pull credentials inside `build` pod as 
+  `hostPath`(similar to what is done for Image Stream Import).
+- Builder image parses node's pull credentials and uses them during build,
+  merging with other pull credentials that may exist on the same namespace.
+
+### Risks and Mitigations
+
+#### Pull credentials inside build pod may be visible to the user
+
+Cluster wise credentials mounted inside the builder pod may be a security risk
+as the user may be able to shell into the pod and copy them. Another possible
+risk is that the user may copy the credentials into a resulting image.
+
+Mitigations:
+
+As far as I verified(and this needs to be once more tested) it is impossible
+to spawn a shell inside the builder pods. I also tried to copy the credentials
+from builder's filesystem into a resulting image but it failed.
+
+#### Pushing to registries using node's credentials
+
+If pull credentials are used during builds we may allow users to push images to
+any registry configured on the node. This is not an ideal scenario as cluster
+administrators may not to be aware or wanting this.
+
+Mitigations
+
+Pull credentials must be used **only** for pulling images. On pushing we must
+use only default credentials'.
+
+#### Absence of mounted path on node filesystem
+
+If the path we are trying to mount through `hostPath` directive does not exist
+on the node where the pod is running the pod won't come up.
+
+Mitigations
+
+Mounted path always exist on master nodes(_still awaiting for node's team check
```

Comment:
we don't just run build pods on master nodes, we run them on worker nodes (in fact i doubt we run them on master nodes at all currently..i do not think we tolerate that taint, but that taint could be removed in the future)

Reply:
fixed.